src/peer_manager.rs

```diff
@@ -1215,24 +1213,34 @@ impl PeerManager {
         let _ = self.set_state(peer_id, PeerState::Routing(RoutingConnection::Direct));
     }
 
-    /// Returns the `NUM_TUNNEL_VIA_NODES` closest peers from our section which can be potential
-    /// tunnel node.
-    pub fn potential_tunnel_nodes(&self) -> Vec<(XorName, PeerId)> {
-        let our_section = self.routing_table
-            .other_close_names(self.our_public_id.name())
-            .unwrap_or_default();
-        self.peer_map
-            .peers()
-            .filter_map(|peer| if our_section.contains(peer.name()) &&
-                                  peer.state.can_tunnel_for() {
-                peer.peer_id.map_or(None, |peer_id| Some((*peer.name(), peer_id)))
-            } else {
-                None
+    /// Returns direct-connected peers in our section and in the peer's section.
+    pub fn potential_tunnel_nodes(&self, name: &XorName) -> Vec<(XorName, PeerId)> {
+        self.routing_table
+            .our_section()
+            .iter()
+            .chain(self.routing_table.get_section(name).unwrap_or(&BTreeSet::new()).iter())
+            .filter_map(|name| {
+                if name == self.our_public_id.name() {
+                    return None;
+                }
+                self.peer_map.get_by_name(name).and_then(|peer| if peer.state.can_tunnel_for() {
+                    peer.peer_id().and_then(|peer_id| Some((*name, *peer_id)))
+                } else {
+                    None
+                })
             })
-            .take(NUM_TUNNEL_VIA_NODES)
             .collect()
     }
 
+    /// Returns true if peer is direct-connected and in our section or in tunnel_client's section.
+    pub fn is_potential_tunnel_node(&self, peer: &PublicId, tunnel_client: &XorName) -> bool {
+        (self.routing_table.our_section().contains(peer.name()) ||
```

Comment:
this function is only being called in node::add_to_routing_table
if peer is ourself, we shall already return at https://github.com/maidsafe/routing/blob/master/src/states/node.rs#L1404-L1410, because routing_table::add will return an error for that case https://github.com/maidsafe/routing/blob/master/src/routing_table/mod.rs#L502-L504 anyway (also I think peer_mgr won't have `peer` info for self?)

so, I think we are safe to not carry out an additional check here?